### DISCUSSION TITLE
Dependency Injection decorator

### DISCUSSION THREAD

**Discussion by tyzhnenko** [Neoteroi/BlackSheep#336] [Ideas]
## Idea

Add a global container for services like _RoutesRegistry_ in server.controllers module for routes. And provide a decorator to fill the container.

### Current situation

Currently, we have to register services via the _Application_ instance. It works well in case we use like it this:

```python

app = Application()

app.services.add_exact_transient(A)
app.services.add_exact_scoped(B)
app.services.add_exact_singleton(C)

app.services.add_exact_scoped(Foo)
```

But when we set up _Application_ right before starting a server like in MVC boilerplate, we should register services somewhere next to configuration functions. It leads to importing classes just to add them to the service container.

### Workaround example

I write a workaround like the _fixture_ decorator in PyTest. Use pre-created container and fill it by the decorator. 

```python
#
# service.py
# 
# Singleton container for services
container = Container()

ScopeType = Literal["scoped", "singleton", "transient"]


def add_service(scope: ScopeType = "scoped"):
    """Decorator to register a service in the container."""

    def deco(target: Any) -> Any:
        match scope:
            case "scoped":
                container.add_scoped(target)
            case "singleton":
                container.add_singleton(target)
            case "transient":
                container.add_transient(target)
            case _:
                raise ValueError(f"Invalid scope: {scope}")

        return target

    return deco


def configure_services(
    configuration: Configuration,
) -> Tuple[Container, ServicesRegistrationContext, Configuration]:
    context = ServicesRegistrationContext()
    container.add_instance(configuration)

    return container, context, configuration

#
# another_module.py
#
from project.core.service import add_service

@add_service(scope="transient")
class A:
    def __init__(self) -> None:
        pass

@add_service(scope="scoped")
class B:
    def __init__(self) -> None:
        pass

@add_service(scope="singletone")
class C:
    def __init__(self) -> None:
        pass

@add_service(scope="scoped")
class Foo:
    def __init__(self, a1: A, a2: A, b1: B, b2: B, c1: C, c2: C) -> None:
        self.a1 = a1
        self.a2 = a2
        self.b1 = b1
        self.b2 = b2
        self.c1 = c1
        self.c2 = c2

```

**Comment by RobertoPrevato** [comment]
@tyzhnenko I love this idea. It also looks compatible in philosophy to https://github.com/Neoteroi/BlackSheep/issues/334

At first sight, I would like a code API like this:
- `@transient()`
- `@scoped()`
- `@singleton()`

```python
from blacksheep.services import transient, scoped, singleton  # <-- names can be like this because the namespace clarifies what they are about (like `from json import dumps`)

@transient
class Foo:
    ...
```

There is a small complication (but we can ignore it, since this feature would require explicit opting-in): I wished to add support for alternative implementations of dependency injection. This is why I introduced the `ContainerProtocol` in `rodi` and added the `blacksheep.settings.di` module. Some other implementations of dependency injections do not implement the concept of scoped service, only transient or singletons. But we can accept that this feature is only available with `rodi`, for example, or think of a way to still support other types of containers.
